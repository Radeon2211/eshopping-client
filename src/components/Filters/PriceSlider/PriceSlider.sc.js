import styled from 'styled-components';

export const Wrapper = styled.div`
  position: relative;

  & .input {
    appearance: none;
    display: inline;
    height: 1rem;
    opacity: 0;
    pointer-events: none;
    position: absolute;
    width: 100%;
    z-index: ${({ theme }) => theme.zIndexes.level2};

    &::-webkit-slider-thumb {
      appearance: none;
      border: 0 none;
      border-radius: 0;
      cursor: pointer;
      height: 2.4rem;
      pointer-events: all;
      width: 2.4rem;
    }
  }
`;

export const Slider = styled.div`
  height: 1rem;
  position: relative;
  z-index: ${({ theme }) => theme.zIndexes.level1};

  & .track {
    background-color: ${({ theme }) => theme.colors.light2};
    border-radius: 5px;
    height: 100%;
    left: 0;
    position: absolute;
    top: 0;
    width: 100%;
    z-index: ${({ theme }) => theme.zIndexes.level1};
  }

  & .range {
    background-color: ${({ theme }) => theme.colors.blue};
    bottom: 0;
    border-radius: 5px;
    left: ${({ positions: { left } }) => left}%;
    position: absolute;
    top: 0;
    right: ${({ positions: { right } }) => right}%;
    z-index: ${({ theme }) => theme.zIndexes.level2};
  }

  & .thumb {
    background-color: ${({ theme }) => theme.colors.light1};
    border: 2px solid ${({ theme }) => theme.colors.blue};
    border-radius: 50%;
    height: 2.4rem;
    position: absolute;
    width: 2.4rem;
    z-index: ${({ theme }) => theme.zIndexes.level3};

    &.left {
      left: ${({ positions: { left } }) => left}%;
      transform: translate(-${({ positions: { left } }) => left}%, -0.7rem);
    }

    &.right {
      right: ${({ positions: { right } }) => right}%;
      transform: translate(${({ positions: { right } }) => right}%, -0.7rem);
    }
  }
`;
