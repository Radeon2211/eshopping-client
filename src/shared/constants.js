import * as Yup from 'yup';
import { getCountries } from 'country-fns';

export const anyStringRule = Yup.string().min(1).required();

export const userRules = {
  email: Yup.string().email().max(320).trim().required(),
  hideEmail: Yup.bool(),
  username: Yup.string().min(3).max(20).trim().required(),
  password: Yup.string().min(7).max(64).trim().required(),
  firstName: Yup.string().max(60).trim().required(),
  lastName: Yup.string().max(80).trim().required(),
  phonePrefix: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
  phoneNumber: Yup.string()
    .min(5)
    .max(15)
    .matches(/^[\d-]{5,15}$/)
    .trim()
    .required(),
  hidePhone: Yup.bool(),
  street: Yup.string().max(60).trim().required(),
  zipCode: Yup.string().max(12).trim().required(),
  city: Yup.string().max(100).trim().required(),
  country: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
};

export const productConditions = {
  NEW: 'new',
  USED: 'used',
  NOT_APPLICABLE: 'not_applicable',
};

export const productRules = {
  name: Yup.string().max(150).trim().required(),
  price: Yup.number().min(0.01).max(1000000).required(),
  quantity: Yup.number().min(1).max(100000).required(),
  condition: Yup.string()
    .oneOf([productConditions.NEW, productConditions.USED, productConditions.NOT_APPLICABLE])
    .required(),
  description: Yup.string().max(600).trim(),
};

export const listOfAreaCodes = getCountries().map(({ name, dial }) => {
  const finalValue = `+${dial} ${name.split('(')[0].trim()}`;
  return {
    value: dial,
    label: finalValue,
  };
});

export const listOfCountries = getCountries().map(({ name }) => {
  const finalValue = name.split('(')[0].trim();
  return {
    value: finalValue,
    label: finalValue,
  };
});

export const productPages = {
  ALL_PRODUCTS: 'ALL_PRODUCTS',
  MY_PRODUCTS: 'MY_PRODUCTS',
  USER_PRODUCTS: 'USER_PRODUCTS',
};

export const itemTypes = {
  CART: 'CART',
  ORDER: 'ORDER',
  TRANSACTION: 'TRANSACTION',
};

export const listItemTypes = {
  PRODUCT: 'PRODUCT',
  ORDER: 'ORDER',
};

export const orderTypes = {
  PLACED_ORDERS: 'PLACED_ORDERS',
  SELL_HISTORY: 'SELL_HISTORY',
};

export const inputKinds = {
  INPUT: 'INPUT',
  TEXTAREA: 'TEXTAREA',
  SELECT: 'SELECT',
  RADIO: 'RADIO',
};

export const defaultProductsPerPage = 10;

export const defaultOrdersPerPage = 6;

export const defaultAppPath = '/products?p=1';

export const filtersActions = {
  INIT_STATE: 'INIT_STATE',
  SET_SORT_BY: 'SET_SORT_BY',
  SET_MIN_PRICE: 'SET_MIN_PRICE',
  SET_MAX_PRICE: 'SET_MAX_PRICE',
  SET_CONDITION: 'SET_CONDITION',
};

export const sliderPositionsActions = {
  SET_BOTH: 'SET_BOTH',
  SET_LEFT: 'SET_LEFT',
  SET_RIGHT: 'SET_RIGHT',
};

export const modalTypes = {
  ABOUT_WEBSITE: 'ABOUT_WEBSITE',
  ADD_ADMIN: 'ADD_ADMIN',
  ADD_PRODUCT: 'ADD_PRODUCT',
  BUY_PRODUCTS: 'BUY_PRODUCTS',
  CART_ITEM_ADDED: 'CART_ITEM_ADDED',
  CHANGE_NAME: 'CHANGE_NAME',
  CHANGE_EMAIL: 'CHANGE_EMAIL',
  CHANGE_USERNAME: 'CHANGE_USERNAME',
  CHANGE_PHONE_NUMBER: 'CHANGE_PHONE_NUMBER',
  CHANGE_ADDRESS: 'CHANGE_ADDRESS',
  CHANGE_PASSWORD: 'CHANGE_PASSWORD',
  CHANGE_CONTACTS: 'CHANGE_CONTACTS',
  CHANGE_DELIVERY_ADDRESS: 'CHANGE_DELIVERY_ADDRESS',
  CLEAR_CART: 'CLEAR_CART',
  DELETE_ACCOUNT: 'DELETE_ACCOUNT',
  DELETE_PRODUCT: 'DELETE_PRODUCT',
  EDIT_PRODUCT: 'EDIT_PRODUCT',
  LOGIN: 'LOGIN',
  REMOVE_ADMIN: 'REMOVE_ADMIN',
  RESET_PASSWORD: 'RESET_PASSWORD',
  PENDING_USER_INFO: 'PENDING_USER_INFO',
  SEND_VERIFICATION_LINK: 'SEND_VERIFICATION_LINK',
  SIGNUP: 'SIGNUP',
};

export const sortProductsOptions = [
  { value: undefined, label: 'Default sorting' },
  { value: 'price:asc', label: `Price - ascending` },
  { value: 'price:desc', label: 'Price - descending' },
  { value: 'name:asc', label: 'Name - A to Z' },
  { value: 'name:desc', label: 'Name - Z to A' },
];

export const sortOrdersOptions = [
  { value: 'createdAt:desc', label: 'Date - from newest' },
  { value: 'createdAt:asc', label: 'Date - from oldest' },
  { value: 'overallPrice:asc', label: `Price - ascending` },
  { value: 'overallPrice:desc', label: 'Price - descending' },
];

export const singleInfoNames = {
  USERNAME: 'Username',
  NAME: 'Name',
  EMAIL: 'Email',
  PHONE_NUMBER: 'Phone number',
  ADDRESS: 'Address',
  CONTACTS: 'Contacts',
};

export const updateCartActions = {
  INCREMENT: 'INCREMENT',
  DECREMENT: 'DECREMENT',
  NUMBER: 'NUMBER',
};

export const defaultErrorMessage = 'Something went wrong';

export const productsPerPageControllerOptions = [
  { value: 1, label: '1' },
  { value: 2, label: '2' },
  { value: 3, label: '3' },
  { value: 4, label: '4' },
  { value: 5, label: '5' },
  { value: 10, label: '10' },
  { value: 15, label: '15' },
  { value: 20, label: '20' },
  { value: 25, label: '25' },
];

export const defaultScrollToConfig = {
  top: 0,
  left: 0,
  behavior: 'smooth',
};

export const userStatuses = {
  ACTIVE: 'active',
  PENDING: 'pending',
};
