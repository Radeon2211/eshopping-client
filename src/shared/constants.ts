import * as Yup from 'yup';
import countryCodes from 'country-codes-list';
import { ProductCondition, ProductPhotoFieldValue } from './types/enums';

export const anyStringRule = Yup.string().min(1).required();

export const userRules = {
  email: Yup.string().email().max(320).trim().required(),
  hideEmail: Yup.bool(),
  username: Yup.string().min(3).max(20).trim().required(),
  password: Yup.string().min(7).max(64).trim().required(),
  firstName: Yup.string().max(60).trim().required(),
  lastName: Yup.string().max(80).trim().required(),
  phonePrefix: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
  phoneNumber: Yup.string()
    .min(5)
    .max(15)
    .matches(/^[\d-]{5,15}$/)
    .trim()
    .required(),
  hidePhone: Yup.bool(),
  street: Yup.string().max(60).trim().required(),
  zipCode: Yup.string().max(12).trim().required(),
  city: Yup.string().max(100).trim().required(),
  country: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
};

export const productRules = {
  name: Yup.string().max(150).trim().required(),
  price: Yup.number().min(0.01).max(1000000).required(),
  quantity: Yup.number().min(1).max(100000).required(),
  condition: Yup.string()
    .oneOf([ProductCondition.NEW, ProductCondition.USED, ProductCondition.NOT_APPLICABLE])
    .required(),
  description: Yup.string().max(600).trim(),
  photo: Yup.mixed()
    .test('is-error', (value) => value !== ProductPhotoFieldValue.ERROR)
    .nullable(),
};

export const listOfAreaCodes = countryCodes.customArray({
  value: '{countryCallingCode}',
  label: '+{countryCallingCode} {countryNameEn}',
});

export const listOfCountries = countryCodes.customArray({
  value: '{countryNameEn}',
  label: '{countryNameEn}',
});

export const itemTypes = {
  CART: 'CART',
  ORDER: 'ORDER',
  TRANSACTION: 'TRANSACTION',
};

export const listItemTypes = {
  PRODUCT: 'PRODUCT',
  ORDER: 'ORDER',
};

export const inputKinds = {
  INPUT: 'INPUT',
  TEXTAREA: 'TEXTAREA',
  SELECT: 'SELECT',
  RADIO: 'RADIO',
};

export const defaultProductsPerPage = 10;

export const defaultOrdersPerPage = 6;

export const defaultAppPath = '/products?p=1';

export const filtersActions = {
  INIT_STATE: 'INIT_STATE',
  SET_SORT_BY: 'SET_SORT_BY',
  SET_MIN_PRICE: 'SET_MIN_PRICE',
  SET_MAX_PRICE: 'SET_MAX_PRICE',
  SET_CONDITION: 'SET_CONDITION',
};

export const sliderPositionsActions = {
  SET_BOTH: 'SET_BOTH',
  SET_LEFT: 'SET_LEFT',
  SET_RIGHT: 'SET_RIGHT',
};

export const sortProductsOptions = [
  { value: undefined, label: 'Default sorting' },
  { value: 'price:asc', label: 'Price - ascending' },
  { value: 'price:desc', label: 'Price - descending' },
  { value: 'name:asc', label: 'Name - A to Z' },
  { value: 'name:desc', label: 'Name - Z to A' },
];

export const sortOrdersOptions = [
  { value: 'createdAt:desc', label: 'Date - from newest' },
  { value: 'createdAt:asc', label: 'Date - from oldest' },
  { value: 'overallPrice:asc', label: 'Price - ascending' },
  { value: 'overallPrice:desc', label: 'Price - descending' },
];

export const singleInfoNames = {
  USERNAME: 'Username',
  NAME: 'Name',
  EMAIL: 'Email',
  PHONE_NUMBER: 'Phone number',
  ADDRESS: 'Address',
  CONTACTS: 'Contacts',
};

export const defaultErrorMessage = 'Something went wrong';

export const productsPerPageControllerOptions = [
  { value: 1, label: '1' },
  { value: 2, label: '2' },
  { value: 3, label: '3' },
  { value: 4, label: '4' },
  { value: 5, label: '5' },
  { value: 10, label: '10' },
  { value: 15, label: '15' },
  { value: 20, label: '20' },
  { value: 25, label: '25' },
];

export const defaultScrollToConfig: ScrollToOptions = {
  top: 0,
  left: 0,
  behavior: 'smooth',
};
