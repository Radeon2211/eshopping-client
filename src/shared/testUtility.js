// eslint-disable-next-line import/no-extraneous-dependencies
import checkPropTypes from 'check-prop-types';

export const checkProps = (component, expectedProps) => {
  // eslint-disable-next-line react/forbid-foreign-prop-types
  return checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

export const historyPageNum = (pageNumber) => ({
  listen: jest.fn(),
  location: { search: `?p=${pageNumber}` },
  createHref: jest.fn(),
  push: jest.fn(),
});

export const propsPagination = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  maxQuantityPerPage: 2,
});
