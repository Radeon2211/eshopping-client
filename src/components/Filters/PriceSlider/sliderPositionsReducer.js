import { updateObject } from '../../../shared/utility/utility';
import { sliderPositionsActions } from '../../../shared/constants';

export const sliderPositionsInitialState = {
  left: 0,
  right: 0,
};

// eslint-disable-next-line default-param-last
export const sliderPositionsReducer = (state = sliderPositionsInitialState, action) => {
  switch (action.type) {
    case sliderPositionsActions.SET_BOTH:
      return { ...action.payload };
    case sliderPositionsActions.SET_LEFT:
      return updateObject(state, { left: action.left });
    case sliderPositionsActions.SET_RIGHT:
      return updateObject(state, { right: action.right });
    default:
      return state;
  }
};
