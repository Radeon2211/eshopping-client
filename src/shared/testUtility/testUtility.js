// eslint-disable-next-line import/no-extraneous-dependencies
import checkPropTypes from 'check-prop-types';
import { v4 as uuidv4 } from 'uuid';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from '../../store/reducers/rootReducer';
import { productConditions } from '../constants';

export const defaultDeliveryAddress = {
  firstName: 'firstName',
  lastName: 'lastName',
  street: 'street',
  zipCode: 'zipCode',
  city: 'city',
  country: 'country',
  phone: 'phone',
};

export const defaultUserProfile = {
  ...defaultDeliveryAddress,
  _id: 'userId',
  username: 'username',
  email: 'email@domain.com',
  status: 'active',
  contacts: {
    email: true,
    phone: true,
  },
  createdAt: '2021-02-10T19:10:38.872Z',
};

export const checkProps = (component, expectedProps) => {
  // eslint-disable-next-line react/forbid-foreign-prop-types
  return checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

export const createHistoryPageNumber = (pageNumber = 1, push = jest.fn()) => ({
  listen: jest.fn(),
  location: { pathname: '/products', search: `?p=${pageNumber}` },
  createHref: jest.fn(),
  push,
});

export const createPaginationProps = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  quantityPerPage: 2,
});

export const createProductItem = (
  productId,
  sellerUsername = 'username',
  quantity = 1,
  price = 2,
  name = 'product name',
  photo = false,
  quantitySold = 0,
  buyerQuantity = 0,
  description = '',
  condition = productConditions.NEW,
) => ({
  _id: productId || uuidv4(),
  name,
  price,
  quantity,
  quantitySold,
  buyerQuantity,
  description,
  condition,
  photo,
  seller: {
    username: sellerUsername,
  },
  __v: 0,
  createdAt: '2021-02-10T19:10:38.872Z',
  updatedAt: '2021-02-10T19:10:38.872Z',
});

export const createCartItem = (
  sellerUsername = 'username',
  quantity = 1,
  productId,
  price = 2,
  name = 'product name',
  productQuantity = 2,
  photo = false,
) => {
  const product = createProductItem(
    productId || uuidv4(),
    sellerUsername,
    productQuantity,
    price,
    name,
    photo,
  );

  return {
    _id: uuidv4(),
    quantity,
    product,
  };
};

export const createTransactionAndOrderProdItem = (
  productId,
  sellerUsername = 'username',
  quantity = 1,
  price = 2,
  name = 'product name',
  photo = false,
) => ({
  _id: productId || uuidv4(),
  name,
  photo,
  price,
  quantity,
  seller: {
    username: sellerUsername,
  },
});

export const createOrder = (
  products = [createTransactionAndOrderProdItem()],
  orderId = uuidv4(),
  sellerUsername,
  buyerUsername,
  overallPrice,
  createdAt,
  sellerEmail,
  sellerPhone,
) => {
  const seller = sellerUsername
    ? {
        username: sellerUsername,
        email: sellerEmail,
        phone: sellerPhone,
      }
    : null;

  const buyer = buyerUsername
    ? {
        username: buyerUsername,
      }
    : null;

  return {
    _id: orderId,
    seller,
    buyer,
    products,
    overallPrice,
    deliveryAddress: defaultDeliveryAddress,
    createdAt,
    updatedAt: createdAt,
    __v: 0,
  };
};

export const testStore = (initialAuth = {}, initialProduct = {}, initialUI = {}) => {
  const createStoreWithMiddleware = applyMiddleware(thunk)(createStore);
  const defaultState = createStoreWithMiddleware(rootReducer).getState();

  const finalState = {
    auth: {
      ...defaultState.auth,
      ...initialAuth,
    },
    product: {
      ...defaultState.product,
      ...initialProduct,
    },
    ui: {
      ...defaultState.ui,
      ...initialUI,
    },
  };
  const finalStore = createStoreWithMiddleware(rootReducer, finalState);

  return {
    store: finalStore,
    initialState: finalStore.getState(),
  };
};

export const checkReqMethodAndURL = (moxiosInstance, method, url, reqNumber = 0) => {
  if (reqNumber) {
    const requests = moxiosInstance.requests.__items;
    const prevRequest = requests[reqNumber - 1];
    return method === prevRequest.config.method && url === prevRequest.config.url;
  }
  const lastRequest = moxiosInstance.requests.mostRecent();
  return method === lastRequest.config.method && url === lastRequest.config.url;
};

export const createExpectedState = (
  initialState,
  expectedAuth = {},
  expectedProduct = {},
  expectedUI = {},
) => {
  const newState = {
    auth: {
      ...initialState.auth,
      ...expectedAuth,
    },
    product: {
      ...initialState.product,
      ...expectedProduct,
    },
    ui: {
      ...initialState.ui,
      ...expectedUI,
    },
  };

  return newState;
};

export const setUpStoreWithDefaultProfile = (auth = {}, product = {}, ui = {}) => {
  return testStore(
    {
      profile: defaultUserProfile,
      deliveryAddress: defaultDeliveryAddress,
      cart: [],
      ...auth,
    },
    product,
    ui,
  );
};
