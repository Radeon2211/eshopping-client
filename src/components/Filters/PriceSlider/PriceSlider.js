import React, { useState, useReducer, useEffect } from 'react';
import PropTypes from 'prop-types';
import queryString from 'query-string';
import { useHistory } from 'react-router-dom';
import { useSelector } from 'react-redux';
import * as SC from './PriceSlider.sc';
import NumberInput from '../../UI/NumberInput/NumberInput';
import PlainText from '../../UI/PlainText';
import { filtersActions, sliderPositionsActions } from '../../../shared/constants';
import { getParamsWithoutPollution } from '../../../shared/utility/utility';
import { sliderPositionsReducer, sliderPositionsInitialState } from './sliderPositionsReducer';

export default function PriceSlider({ dispatchFilters }) {
  const history = useHistory();
  const {
    location: { search },
  } = history;

  const minPriceState = useSelector((state) => state.product.minPrice);
  const maxPriceState = useSelector((state) => state.product.maxPrice);

  const [positions, dispatchPositions] = useReducer(
    sliderPositionsReducer,
    sliderPositionsInitialState,
  );
  const [inputValues, setInputValues] = useState({
    minPrice: 0,
    maxPrice: 0,
  });
  const [rangeValues, setRangeValues] = useState({
    minPrice: 0,
    maxPrice: 0,
  });

  useEffect(() => {
    if (minPriceState === 0 || maxPriceState === 0) return;

    const correctQueryParams = getParamsWithoutPollution(search);

    const { minPrice, maxPrice } = correctQueryParams;
    const queryMinPrice = Number.parseFloat((+minPrice).toFixed(2));
    let queryMaxPrice = Number.parseFloat((+maxPrice).toFixed(2));

    let changeQueryParams = false;
    if (queryMaxPrice && (queryMaxPrice > maxPriceState || queryMaxPrice <= minPriceState)) {
      queryMaxPrice = maxPriceState;
      delete correctQueryParams.maxPrice;
      changeQueryParams = true;
    }
    if (
      queryMinPrice &&
      (queryMinPrice >= (queryMaxPrice || maxPriceState) || queryMinPrice < minPriceState)
    ) {
      delete correctQueryParams.minPrice;
      changeQueryParams = true;
    }
    if (changeQueryParams) {
      const stringifiedQueryParams = queryString.stringify(correctQueryParams);
      history.replace(`${history.location.pathname}?${stringifiedQueryParams}`);
      return;
    }

    const minValueToCalculate = (queryMinPrice || minPriceState) - minPriceState;
    const maxValueToCalculate = (queryMaxPrice || maxPriceState) - minPriceState;
    const maxPriceStateToCalculate = maxPriceState - minPriceState;

    let leftPercentValue = 0;
    let rightPercentValue = 0;
    if (maxPriceStateToCalculate > 0) {
      leftPercentValue = (100 / maxPriceStateToCalculate) * minValueToCalculate;
      rightPercentValue = 100 - (100 / maxPriceStateToCalculate) * maxValueToCalculate;
      if (leftPercentValue < 0) leftPercentValue = 0;
      if (leftPercentValue > 100) leftPercentValue = 100;
      if (rightPercentValue < 0) rightPercentValue = 0;
      if (rightPercentValue > 100) rightPercentValue = 100;
    }

    dispatchPositions({
      type: sliderPositionsActions.SET_BOTH,
      payload: { left: leftPercentValue, right: rightPercentValue },
    });
    setInputValues({
      minPrice: queryMinPrice || minPriceState,
      maxPrice: queryMaxPrice || maxPriceState,
    });
    setRangeValues({
      minPrice: queryMinPrice || minPriceState,
      maxPrice: queryMaxPrice || maxPriceState,
    });
    dispatchFilters({
      type: filtersActions.SET_MIN_PRICE,
      minPrice: queryMinPrice || minPriceState,
    });
    dispatchFilters({
      type: filtersActions.SET_MAX_PRICE,
      maxPrice: queryMaxPrice || maxPriceState,
    });
  }, [minPriceState, maxPriceState, setRangeValues, dispatchFilters, history, search]);

  const inputChangeHandle = (e) => {
    e.persist();
    setInputValues((prevState) => ({
      ...prevState,
      [e.target.name]: +e.target.value || '',
    }));
  };

  const validateInputValue = (e) => {
    e.persist();

    let value = Number.parseFloat((+e.target.value).toFixed(2));

    const maxPriceStateToCalculate = maxPriceState - minPriceState;
    let percentValue = 0;
    if (e.target.name === 'minPrice') {
      if (value < minPriceState) {
        value = minPriceState;
      } else if (value > rangeValues.maxPrice) {
        value = rangeValues.maxPrice;
      }
      const minValueToCalculate = (value || minPriceState) - minPriceState;
      if (maxPriceStateToCalculate !== 0) {
        percentValue = (100 / maxPriceStateToCalculate) * minValueToCalculate;
      }
      dispatchPositions({ type: sliderPositionsActions.SET_LEFT, left: percentValue });
      dispatchFilters({ type: filtersActions.SET_MIN_PRICE, minPrice: value });
    } else {
      if (value > maxPriceState) {
        value = maxPriceState;
      } else if (value < rangeValues.minPrice) {
        value = rangeValues.minPrice;
      }
      const maxValueToCalculate = (value || maxPriceState) - minPriceState;
      if (maxPriceStateToCalculate !== 0) {
        percentValue = 100 - (100 / maxPriceStateToCalculate) * maxValueToCalculate;
      }
      dispatchPositions({ type: sliderPositionsActions.SET_RIGHT, right: percentValue });
      dispatchFilters({ type: filtersActions.SET_MAX_PRICE, maxPrice: value });
    }

    setInputValues((prevState) => ({
      ...prevState,
      [e.target.name]: value,
    }));
    setRangeValues((prevState) => ({
      ...prevState,
      [e.target.name]: value,
    }));
  };

  const rangeChangeHandle = (e) => {
    e.persist();

    let parsedValue = Number.parseFloat((+e.target.value).toFixed(2));

    if (
      (e.target.name === 'minPrice' && parsedValue >= rangeValues.maxPrice) ||
      (e.target.name === 'maxPrice' && parsedValue <= rangeValues.minPrice)
    ) {
      return;
    }

    const maxPriceStateToCalculate = maxPriceState - minPriceState;
    let percentValue = 0;
    if (e.target.name === 'minPrice') {
      if (parsedValue - minPriceState <= maxPriceState / 1000) {
        parsedValue = minPriceState;
      }
      const minValueToCalculate = (parsedValue || minPriceState) - minPriceState;
      if (maxPriceStateToCalculate !== 0) {
        percentValue = (100 / maxPriceStateToCalculate) * minValueToCalculate;
      }
      dispatchPositions({ type: sliderPositionsActions.SET_LEFT, left: percentValue });
      dispatchFilters({ type: filtersActions.SET_MIN_PRICE, minPrice: parsedValue });
    } else {
      if (maxPriceState - parsedValue <= maxPriceState / 1000) {
        parsedValue = maxPriceState;
      }
      const maxValueToCalculate = (parsedValue || maxPriceState) - minPriceState;
      if (maxPriceStateToCalculate !== 0) {
        percentValue = 100 - (100 / maxPriceStateToCalculate) * maxValueToCalculate;
      }
      dispatchPositions({ type: sliderPositionsActions.SET_RIGHT, right: percentValue });
      dispatchFilters({ type: filtersActions.SET_MAX_PRICE, maxPrice: parsedValue });
    }

    setInputValues((prevState) => ({
      ...prevState,
      [e.target.name]: parsedValue,
    }));
    setRangeValues((prevState) => ({
      ...prevState,
      [e.target.name]: parsedValue,
    }));
  };

  return (
    <SC.Wrapper>
      <SC.LabelAndInputs>
        <PlainText size="3">Price ($)</PlainText>
        <div className="inputs">
          <NumberInput
            name="minPrice"
            size="big"
            value={inputValues.minPrice}
            changed={inputChangeHandle}
            blured={validateInputValue}
            floating
          />
          <span className="inputs-gap">&mdash;</span>
          <NumberInput
            name="maxPrice"
            size="big"
            value={inputValues.maxPrice}
            changed={inputChangeHandle}
            blured={validateInputValue}
            floating
          />
        </div>
      </SC.LabelAndInputs>
      <SC.Slider positions={positions}>
        <input
          type="range"
          name="minPrice"
          min={minPriceState}
          max={maxPriceState}
          step={maxPriceState / 1000}
          value={rangeValues.minPrice}
          onChange={rangeChangeHandle}
          className="input min"
          data-testid="PriceSlider-price-range-min"
        />
        <input
          type="range"
          name="maxPrice"
          min={minPriceState}
          max={maxPriceState}
          step={maxPriceState / 1000}
          value={rangeValues.maxPrice}
          onChange={rangeChangeHandle}
          className="input max"
          data-testid="PriceSlider-price-range-max"
        />
        <div className="track" />
        <div className="range" />
        <div className="thumb left" />
        <div className="thumb right" />
      </SC.Slider>
    </SC.Wrapper>
  );
}

PriceSlider.propTypes = {
  dispatchFilters: PropTypes.func.isRequired,
};
