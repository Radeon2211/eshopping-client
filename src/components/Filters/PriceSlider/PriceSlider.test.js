/* eslint-disable no-console */
import React from 'react';
import { mount } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import thunk from 'redux-thunk';
import PriceSlider from './PriceSlider';
import { checkProps } from '../../../shared/testUtility';
import theme from '../../../styled/theme';
import { sliderPositionsActions } from '../../../shared/constants';
import { sliderPositionsReducer, sliderPositionsInitialState } from './sliderPositionsReducer';

const mockStore = configureMockStore([thunk]);

const defaultStore = mockStore({
  product: {
    minPrice: 10,
    maxPrice: 100,
  },
});

const defaultProps = {
  dispatchFilters: jest.fn(),
};

const setUp = (search = '', replace = jest.fn()) => {
  const history = {
    listen: jest.fn(),
    createHref: jest.fn(),
    location: { pathname: '/products', search },
    replace,
  };
  return mount(
    <Router history={history}>
      <Provider store={defaultStore}>
        <ThemeProvider theme={theme}>
          <PriceSlider {...defaultProps} />
        </ThemeProvider>
      </Provider>
    </Router>,
  );
};

describe('<PriceSlider />', () => {
  describe('Check prop types', () => {
    it('Should NOT throw a warning', () => {
      expect(checkProps(PriceSlider, defaultProps)).toBeUndefined();
    });

    it('Should throw a warning', () => {
      expect(checkProps(PriceSlider, {})).not.toBe(null);
    });
  });

  describe(`Check if inputs' values are correct`, () => {
    it('Should min inputs have 10 and max inputs 100 by default', () => {
      const wrapper = setUp();
      expect(wrapper.find('[data-test="min-price-input"]').prop('value')).toBe(10);
      expect(wrapper.find('[data-test="min-price-range"]').prop('value')).toBe(10);
      expect(wrapper.find('[data-test="max-price-input"]').prop('value')).toBe(100);
      expect(wrapper.find('[data-test="max-price-range"]').prop('value')).toBe(100);
    });

    it('Should min inputs have 20 and max inputs 80 when these are set in url', () => {
      const wrapper = setUp('?minPrice=20&maxPrice=80');
      expect(wrapper.find('[data-test="min-price-input"]').prop('value')).toBe(20);
      expect(wrapper.find('[data-test="min-price-range"]').prop('value')).toBe(20);
      expect(wrapper.find('[data-test="max-price-input"]').prop('value')).toBe(80);
      expect(wrapper.find('[data-test="max-price-range"]').prop('value')).toBe(80);
    });

    it('Should call history.replace() when lower and higher than default are set in url', () => {
      const replaceFn = jest.fn();
      setUp('?p=1&minPrice=5&maxPrice=120', replaceFn);
      expect(replaceFn).toHaveBeenCalledWith('/products?p=1');
    });
  });

  describe('Check sliderPositionsReducer', () => {
    it('Should return default state', () => {
      expect(sliderPositionsReducer(undefined, {})).toEqual(sliderPositionsInitialState);
    });

    it('Should return new state after SET_BOTH', () => {
      const payload = {
        left: 10,
        right: 20,
      };
      expect(
        sliderPositionsReducer(undefined, {
          type: sliderPositionsActions.SET_BOTH,
          payload,
        }),
      ).toEqual(payload);
    });

    it('Should return new state after SET_LEFT', () => {
      expect(
        sliderPositionsReducer(undefined, {
          type: sliderPositionsActions.SET_LEFT,
          left: 10,
        }),
      ).toEqual({
        ...sliderPositionsInitialState,
        left: 10,
      });
    });

    it('Should return new state after SET_RIGHT', () => {
      expect(
        sliderPositionsReducer(undefined, {
          type: sliderPositionsActions.SET_RIGHT,
          right: 20,
        }),
      ).toEqual({
        ...sliderPositionsInitialState,
        right: 20,
      });
    });
  });
});
