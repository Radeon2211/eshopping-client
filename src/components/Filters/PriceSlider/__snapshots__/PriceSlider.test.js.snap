// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PriceSlider /> Check how renders Should render everything correctly with default values 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm iFRHBp"
  >
    <div
      class="sc-AxirZ cGSJER"
    >
      <span
        class="sc-AxhUy eKEkiW"
        color=""
        display="inline-block"
        overflow="visible"
        size="3"
        spacing="0px"
        transform=""
      >
        Price ($)
      </span>
      <div
        class="inputs"
      >
        <input
          class="sc-AxhCb hkYbfm"
          data-testid="NumberInput-minPrice"
          name="minPrice"
          type="number"
          value="10"
        />
        <span
          class="inputs-gap"
        >
          â€”
        </span>
        <input
          class="sc-AxhCb hkYbfm"
          data-testid="NumberInput-maxPrice"
          name="maxPrice"
          type="number"
          value="100"
        />
      </div>
    </div>
    <div
      class="sc-AxiKw gOljnE"
    >
      <input
        class="input min"
        data-testid="PriceSlider-price-range-min"
        max="100"
        min="10"
        name="minPrice"
        step="0.1"
        type="range"
        value="10"
      />
      <input
        class="input max"
        data-testid="PriceSlider-price-range-max"
        max="100"
        min="10"
        name="maxPrice"
        step="0.1"
        type="range"
        value="100"
      />
      <div
        class="track"
      />
      <div
        class="range"
      />
      <div
        class="thumb left"
      />
      <div
        class="thumb right"
      />
    </div>
  </div>
</DocumentFragment>
`;
